File: backend/src/projects/routes/handlers/detail/model_routing.rs
//! `GET` / `PATCH` / `POST resolve` for project Studio step model routing.

use axum::{
    extract::{Path, State},
    http::HeaderMap,
    Json,
};
use uuid::Uuid;

use crate::auth::require_user_uuid;
use crate::error::ApiError;
use crate::projects::model_routing::store::{
    load_project_model_routing, merge_routing_patch, persist_model_routing, ProjectModelDefaults,
};
use crate::projects::model_routing::{resolve_all_effective, resolve_model_id};
use crate::projects::model_routing::{
    ModelRoutingDefaultsBody, ModelRoutingEffectiveEntry, ModelRoutingPatchBody,
    ModelRoutingResolveBody, ModelRoutingResolveResponse, ModelRoutingResponse, ModelSlot,
    ResolveInput, StudioStepSlug,
};
use crate::state::AppState;

use super::super::super::common::{
    require_project_workspace_member_scope, require_project_write_scope,
};

fn defaults_body(d: &ProjectModelDefaults) -> ModelRoutingDefaultsBody {
    ModelRoutingDefaultsBody {
        text_model: d.text_model.clone(),
        multimodal_model: d.multimodal_model.clone(),
        image_model: d.image_model.clone(),
        video_model: d.video_model.clone(),
        voice_model: d.voice_model.clone(),
    }
}

fn normalize_optional_string(v: &Option<String>) -> Option<String> {
    v.as_ref().and_then(|s| {
        let t = s.trim();
        if t.is_empty() {
            None
        } else {
            Some(t.to_string())
        }
    })
}

fn apply_defaults_patch(
    base: &ProjectModelDefaults,
    patch: &crate::projects::model_routing::ModelRoutingDefaultsPatch,
) -> ProjectModelDefaults {
    let mut row = base.clone();
    if let Some(v) = &patch.text_model {
        row.text_model = normalize_optional_string(v);
    }
    if let Some(v) = &patch.multimodal_model {
        row.multimodal_model = normalize_optional_string(v);
    }
    if let Some(v) = &patch.image_model {
        row.image_model = normalize_optional_string(v);
    }
    if let Some(v) = &patch.video_model {
        row.video_model = normalize_optional_string(v);
    }
    if let Some(v) = &patch.voice_model {
        row.voice_model = normalize_optional_string(v);
    }
    row
}

/// `GET /api/v1/projects/{project_id}/model-routing`
#[utoipa::path(
    get,
    path = "/api/v1/projects/{project_id}/model-routing",
    operation_id = "getProjectModelRoutingV1",
    tag = "projects",
    params(("project_id" = Uuid, Path, description = "Project UUID")),
    responses(
        (status = 200, description = "OK", body = ModelRoutingResponse),
        (status = 401, description = "Unauthorized", body = crate::error::ErrorBody),
        (status = 403, description = "Forbidden", body = crate::error::ErrorBody),
        (status = 404, description = "Not found", body = crate::error::ErrorBody)
    ),
    security(("bearerAuth" = []))
)]
pub(crate) async fn get_project_model_routing(
    State(state): State<AppState>,
    Path(project_id): Path<Uuid>,
    headers: HeaderMap,
) -> Result<Json<ModelRoutingResponse>, ApiError> {
    let uid = require_user_uuid(&state, &headers)?;
    let pool = state.require_pool()?;
    let _scope = require_project_workspace_member_scope(&state, uid, project_id).await?;

    let loaded = load_project_model_routing(pool, project_id).await?;
    let effective = resolve_all_effective(pool, uid, project_id, &state).await?;
    Ok(Json(ModelRoutingResponse {
        project_id,
        defaults: defaults_body(&loaded.defaults),
        steps: loaded.routing.steps.clone(),
        effective: effective
            .into_iter()
            .map(|r| ModelRoutingEffectiveEntry {
                step: r.step,
                slot: r.slot,
                model_id: r.model_id,
                source: r.source,
            })
            .collect(),
    }))
}

/// `PATCH /api/v1/projects/{project_id}/model-routing`
#[utoipa::path(
    patch,
    path = "/api/v1/projects/{project_id}/model-routing",
    operation_id = "patchProjectModelRoutingV1",
    tag = "projects",
    params(("project_id" = Uuid, Path, description = "Project UUID")),
    request_body = ModelRoutingPatchBody,
    responses(
        (status = 200, description = "OK", body = ModelRoutingResponse),
        (status = 400, description = "Bad request", body = crate::error::ErrorBody),
        (status = 401, description = "Unauthorized", body = crate::error::ErrorBody),
        (status = 403, description = "Forbidden", body = crate::error::ErrorBody),
        (status = 404, description = "Not found", body = crate::error::ErrorBody)
    ),
    security(("bearerAuth" = []))
)]
pub(crate) async fn patch_project_model_routing(
    State(state): State<AppState>,
    Path(project_id): Path<Uuid>,
    headers: HeaderMap,
    Json(body): Json<ModelRoutingPatchBody>,
) -> Result<Json<ModelRoutingResponse>, ApiError> {
    let uid = require_user_uuid(&state, &headers)?;
    let pool = state.require_pool()?;
    let _scope = require_project_write_scope(&state, uid, project_id).await?;

    if body.defaults.is_none() && body.steps.is_none() {
        return Err(ApiError::BadRequest(
            "expected at least one of defaults or steps".into(),
        ));
    }

    let loaded = load_project_model_routing(pool, project_id).await?;

    if let Some(ref steps) = body.steps {
        for (step, slots) in steps {
            StudioStepSlug::parse(step)
                .ok_or_else(|| ApiError::BadRequest(format!("unknown studio step: {step}")))?;
            for (slot, model_id) in slots {
                ModelSlot::parse(slot)
                    .ok_or_else(|| ApiError::BadRequest(format!("unknown model slot: {slot}")))?;
                if !model_id.trim().is_empty() {
                    crate::vendor::catalog::lookup_detail(model_id.trim(), false).ok_or_else(
                        || ApiError::BadRequest(format!("unknown model_id: {model_id}")),
                    )?;
                }
            }
        }
    }

    let defaults_patch_merged = body
        .defaults
        .as_ref()
        .map(|p| apply_defaults_patch(&loaded.defaults, p));

    let (metadata, defaults, _routing) = merge_routing_patch(
        loaded.metadata_raw.clone(),
        defaults_patch_merged,
        body.steps.clone(),
        &loaded.defaults,
        &loaded.routing,
    );

    persist_model_routing(pool, project_id, &metadata, &defaults).await?;

    get_project_model_routing(State(state), Path(project_id), headers).await
}

/// `POST /api/v1/projects/{project_id}/model-routing/resolve`
#[utoipa::path(
    post,
    path = "/api/v1/projects/{project_id}/model-routing/resolve",
    operation_id = "resolveProjectModelRoutingV1",
    tag = "projects",
    request_body = ModelRoutingResolveBody,
    responses(
        (status = 200, description = "OK", body = ModelRoutingResolveResponse),
        (status = 400, description = "Bad request", body = crate::error::ErrorBody),
        (status = 401, description = "Unauthorized", body = crate::error::ErrorBody),
        (status = 403, description = "Forbidden", body = crate::error::ErrorBody),
        (status = 404, description = "Not found", body = crate::error::ErrorBody)
    ),
    security(("bearerAuth" = []))
)]
pub(crate) async fn resolve_project_model_routing(
    State(state): State<AppState>,
    Path(project_id): Path<Uuid>,
    headers: HeaderMap,
    Json(body): Json<ModelRoutingResolveBody>,
) -> Result<Json<ModelRoutingResolveResponse>, ApiError> {
    let uid = require_user_uuid(&state, &headers)?;
    let pool = state.require_pool()?;
    let _scope = require_project_workspace_member_scope(&state, uid, project_id).await?;

    let step = StudioStepSlug::parse(&body.step)
        .ok_or_else(|| ApiError::BadRequest(format!("unknown studio step: {}", body.step)))?;
    let slot = ModelSlot::parse(&body.slot)
        .ok_or_else(|| ApiError::BadRequest(format!("unknown model slot: {}", body.slot)))?;

    let resolved = resolve_model_id(ResolveInput {
        state: &state,
        pool,
        actor_user_id: uid,
        project_id: Some(project_id),
        step,
        slot,
        request_override: body.model_id.as_deref(),
    })
    .await?;

    Ok(Json(ModelRoutingResolveResponse {
        model_id: resolved.model_id,
        step: resolved.step,
        slot: resolved.slot,
        source: resolved.source,
    }))
}

File: frontend/lib/rust_api/project/model_routing.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../../config.dart';
import '../core.dart';

class ProjectModelRoutingDefaults {
  const ProjectModelRoutingDefaults({
    this.textModel,
    this.multimodalModel,
    this.imageModel,
    this.videoModel,
    this.voiceModel,
  });

  final String? textModel;
  final String? multimodalModel;
  final String? imageModel;
  final String? videoModel;
  final String? voiceModel;

  factory ProjectModelRoutingDefaults.fromJson(Map<String, dynamic> json) {
    return ProjectModelRoutingDefaults(
      textModel: json['text_model'] as String?,
      multimodalModel: json['multimodal_model'] as String?,
      imageModel: json['image_model'] as String?,
      videoModel: json['video_model'] as String?,
      voiceModel: json['voice_model'] as String?,
    );
  }

  Map<String, dynamic> toJson() => <String, dynamic>{
    if (textModel != null) 'text_model': textModel,
    if (multimodalModel != null) 'multimodal_model': multimodalModel,
    if (imageModel != null) 'image_model': imageModel,
    if (videoModel != null) 'video_model': videoModel,
    if (voiceModel != null) 'voice_model': voiceModel,
  };
}

class ModelRoutingEffectiveEntry {
  const ModelRoutingEffectiveEntry({
    required this.step,
    required this.slot,
    required this.modelId,
    required this.source,
  });

  final String step;
  final String slot;
  final String modelId;
  final String source;

  factory ModelRoutingEffectiveEntry.fromJson(Map<String, dynamic> json) {
    return ModelRoutingEffectiveEntry(
      step: json['step'] as String,
      slot: json['slot'] as String,
      modelId: json['model_id'] as String,
      source: json['source'] as String,
    );
  }
}

class ProjectModelRoutingResponse {
  const ProjectModelRoutingResponse({
    required this.projectId,
    required this.defaults,
    required this.steps,
    required this.effective,
  });

  final String projectId;
  final ProjectModelRoutingDefaults defaults;
  final Map<String, Map<String, String>> steps;
  final List<ModelRoutingEffectiveEntry> effective;

  factory ProjectModelRoutingResponse.fromJson(Map<String, dynamic> json) {
    final stepsRaw = json['steps'] as Map<String, dynamic>? ?? const {};
    final steps = <String, Map<String, String>>{};
    for (final entry in stepsRaw.entries) {
      final slotMap = entry.value as Map<String, dynamic>? ?? const {};
      final normalized = <String, String>{};
      for (final slotEntry in slotMap.entries) {
        final raw = slotEntry.value;
        if (raw == null) continue;
        final value = raw.toString().trim();
        if (value.isEmpty) continue;
        normalized[slotEntry.key] = value;
      }
      if (normalized.isNotEmpty) {
        steps[entry.key] = normalized;
      }
    }
    final effectiveList = (json['effective'] as List<dynamic>? ?? const [])
        .map(
          (e) => ModelRoutingEffectiveEntry.fromJson(e as Map<String, dynamic>),
        )
        .toList(growable: false);
    return ProjectModelRoutingResponse(
      projectId: json['project_id'] as String,
      defaults: ProjectModelRoutingDefaults.fromJson(
        json['defaults'] as Map<String, dynamic>,
      ),
      steps: steps,
      effective: effectiveList,
    );
  }

  String? effectiveModelFor({required String step, required String slot}) {
    for (final e in effective) {
      if (e.step == step && e.slot == slot) return e.modelId;
    }
    return null;
  }
}

class ResolvedProjectModel {
  const ResolvedProjectModel({
    required this.modelId,
    required this.step,
    required this.slot,
    required this.source,
  });

  final String modelId;
  final String step;
  final String slot;
  final String source;

  factory ResolvedProjectModel.fromJson(Map<String, dynamic> json) {
    return ResolvedProjectModel(
      modelId: json['model_id'] as String,
      step: json['step'] as String,
      slot: json['slot'] as String,
      source: json['source'] as String,
    );
  }
}

/// `GET /api/v1/projects/{project_id}/model-routing`
Future<ProjectModelRoutingResponse> fetchProjectModelRoutingV1(
  String accessToken,
  String projectId,
) async {
  final uri = Uri.parse(
    '$kApiBaseUrl/api/v1/projects/$projectId/model-routing',
  );
  final res = await http
      .get(uri, headers: rustApiAuthHeaders(accessToken))
      .timeout(const Duration(seconds: 15));
  ensureHttpSuccess(res);
  return ProjectModelRoutingResponse.fromJson(
    jsonDecode(res.body) as Map<String, dynamic>,
  );
}

/// `PATCH /api/v1/projects/{project_id}/model-routing`
Future<ProjectModelRoutingResponse> patchProjectModelRoutingV1(
  String accessToken,
  String projectId, {
  Map<String, dynamic>? defaults,
  Map<String, Map<String, String>>? steps,
}) async {
  final body = <String, dynamic>{};
  if (defaults != null) body['defaults'] = defaults;
  if (steps != null) body['steps'] = steps;
  final uri = Uri.parse(
    '$kApiBaseUrl/api/v1/projects/$projectId/model-routing',
  );
  final res = await http
      .patch(
        uri,
        headers: rustApiJsonAuthHeaders(accessToken),
        body: jsonEncode(body),
      )
      .timeout(const Duration(seconds: 15));
  ensureHttpSuccess(res);
  return ProjectModelRoutingResponse.fromJson(
    jsonDecode(res.body) as Map<String, dynamic>,
  );
}

/// `POST /api/v1/projects/{project_id}/model-routing/resolve`
Future<ResolvedProjectModel> resolveProjectModelV1(
  String accessToken,
  String projectId, {
  required String step,
  required String slot,
  String? modelId,
}) async {
  final uri = Uri.parse(
    '$kApiBaseUrl/api/v1/projects/$projectId/model-routing/resolve',
  );
  final res = await http
      .post(
        uri,
        headers: rustApiJsonAuthHeaders(accessToken),
        body: jsonEncode(<String, dynamic>{
          'step': step,
          'slot': slot,
          if (modelId != null && modelId.isNotEmpty) 'model_id': modelId,
        }),
      )
      .timeout(const Duration(seconds: 15));
  ensureHttpSuccess(res);
  return ResolvedProjectModel.fromJson(
    jsonDecode(res.body) as Map<String, dynamic>,
  );
}
